[17.0][IMP] auth_oidc: add groups' handling

OdyX opened the pull request:
This allows groups' handling via a token's attributes as passed by either a Keycloak or a Gitlab instance serving as IdP.

* This is the 17.0-version of #350, then #372 , by @26houseRobot and @hbrunn , with just minor fixes

@sbidoul : I'd be happy to make any necessary changes!

hbrunn committed:
[IMP] auth_oidc: allow assign groups from token claims

hbrunn commented:
the v14 PR is based on my v12 PR which was merged - why didn't you just forward port this?

OdyX commented:
@hbrunn thanks for asking! As I'm quite fresh in the Odoo ecosystem, I did not see the v12 PR. Care to share a link?

As you can see from the code, my patch works a bit differently; as it appeared that what I needed for group mapping was directly in the access token, there's no usage of the data_endpoint. But I'm also likely not fluent enough in OAuth2 to know if that is really a correct way too.

Well; in any case, I'm happy to work towards merging either this or your v12 PR (or a mix of both) for v17. We need @sbidoul 's input, right?

hbrunn commented:
you find the v12 PR [here](https://github.com/OCA/server-auth/pull/350)

OdyX committed:
[IMP] auth_oidc: use Command.link() / .unlink() instead of raw numbers

Co-authored-by: Holger Brunn <mail@hunki-enterprises.com>

OdyX committed:
[IMP] auth_oidc: add oauth.provider.group_line _description attribute

OdyX commented:
@hbrunn Great. Thanks for the pointer to the v12 PR. I've now understood the code much better, and did a mostly-straightforward port, with just two minor additions as separate commits. Could you perhaps review?

OdyX committed:
[IMP] auth_oidc: use raise … from … in exception handling

OdyX committed:
[IMP] auth_oidc: use a 10 seconds' timeout for the GET to data_endpoint

OdyX committed:
[IMP] auth_oidc: parametrize KEYCLOAK_URL in tests

OdyX committed:
[IMP] auth_oidc: test possible token values more extensively

OdyX committed:
[IMP] auth_oidc: test all possible safe_eval exceptions

OdyX committed:
[IMP] auth_oidc: clarify test comments

OdyX commented:
As the codecov warnings seem critical, I've now added some more tests around the safe_eval call of the expressions.

Edit: and now also added some groups' assignment/deassignment checks, pushing the codecov bar above the needed limits.

OdyX committed:
[IMP] auth_oidc: test group assignment/removal

hbrunn commented on auth_oidc/models/res_users.py:
```diff
@@ -80,3 +88,22 @@ def auth_oauth(self, provider, params):
+            ):
+                if group_line._eval_expression(user, validation):
+                    if group_line.group_id not in user.groups_id:
+                        group_updates.append((Command.LINK, group_line.group_id.id))
```
```suggestion
                        group_updates.append(Command.link(group_line.group_id.id))
```

seems more canonical to me

hbrunn commented on auth_oidc/models/res_users.py:
```diff
@@ -80,3 +88,22 @@ def auth_oauth(self, provider, params):
+                        group_updates.append((Command.LINK, group_line.group_id.id))
+                else:
+                    if group_line.group_id in user.groups_id:
+                        group_updates.append((Command.UNLINK, group_line.group_id.id))
```
```suggestion
                        group_updates.append(Command.unlink(group_line.group_id.id))
```

hbrunn reviewed (approved):
thanks, just minor stylistic things

hbrunn commented:
@OdyX please rebase your branch. If you allow edits by maintainers, I could say

/ocabot rebase

and it would do this automatically

OdyX commented:
@hbrunn integrated your suggestions in https://github.com/OCA/server-auth/pull/682/commits/afd10ea74a6ad31bcb60c1b258fab2da95bc65e5; and I have rebased this on top of 17.0.

OdyX commented:
Rebased and pushed. It would be nice to have @sbidoul 's input and get this merged (or commented upon).

sbidoul commented on auth_oidc/models/res_users.py:
```diff
@@ -64,6 +65,13 @@ def auth_oauth(self, provider, params):
+                oauth_provider.data_endpoint,
+                headers={"Authorization": "Bearer %s" % access_token},
+                timeout=10,
+            ).json()
```
Error handling for this request is missing?

sbidoul replied:
I see there is a _auth_oauth_validate in auth_oauth. Does it make sense to reuse it / get inspiration from it ?

OdyX replied:
I've pushed a refactoring of this part towards an _auth_oauth_validate function override. It's not possible to reuse it as-is, but at least it becomes really similar

sbidoul commented on auth_oidc/models/res_users.py:
```diff
@@ -80,3 +88,22 @@ def auth_oauth(self, provider, params):
+                        group_updates.append(Command.unlink(group_line.group_id.id))
+            if group_updates:
+                user.write({"groups_id": group_updates})
+        return login
```
For comparison, do we have other modules that do a similar thing ?

hbrunn replied:
https://github.com/OCA/server-auth/tree/16.0/users_ldap_groups

sbidoul reviewed (commented):
Sorry folks for the delay.

I've not had time to do a deep review yet, but I have a couple a questions from a first read.

OdyX commented:
@sbidoul Could you perhaps take another look?

sbidoul commented on auth_oidc/models/auth_oauth_provider.py:
```diff
@@ -108,3 +114,39 @@ def _decode_id_token(self, access_token, id_token, kid):
+    def _eval_expression(self, user, token):
+        self.ensure_one()
+
+        class Defaultdict2(collections.defaultdict):
```
This comment that @hbrunn added in the original PR was helpful:

```suggestion
        class Defaultdict2(collections.defaultdict):
```
```suggestion
        class Defaultdict2(collections.defaultdict):
            """Class keeping yielding defaultdicts"""
```

sbidoul commented on auth_oidc/models/res_users.py:
```diff
@@ -44,6 +45,45 @@ def _auth_oauth_get_tokens_auth_code_flow(self, oauth_provider, params):
         return response_json.get("access_token"), response_json.get("id_token")
 
+    @api.model
+    def _auth_oauth_validate(self, provider, id_token, access_token):
```
Does this completely replace the upstream `_auth_oauth_validate`? Why is this necessary?

OdyX replied:
Yes. It's mostly a copy-paste from https://github.com/odoo/odoo/blob/17.0/addons/auth_oauth/models/res_users.py#L47-L72 here's the difference:

```diff
--- a/auth_oidc/models/res_users.py
+++ b/auth_oidc/models/res_users.py
@@ -46,16 +46,19 @@ class ResUsers(models.Model):
         return response_json.get("access_token"), response_json.get("id_token")
 
     @api.model
-    def _auth_oauth_validate(self, provider, access_token):
-        """return the validation data corresponding to the access token"""
+    def _auth_oauth_validate(self, provider, id_token, access_token):
+        """
+        return the validation data corresponding to the access token
+        Mostly the same as auth_oauth ResUsers._auth_oauth_validate, minus the
+        validation_endpoint
+        """
         oauth_provider = self.env["auth.oauth.provider"].browse(provider)
-        validation = self._auth_oauth_rpc(
-            oauth_provider.validation_endpoint, access_token
-        )
-        if validation.get("error"):
-            raise Exception(validation["error"])
+
+        # Parse the token to get validation data
+        validation = oauth_provider._parse_id_token(id_token, access_token)
+
         if oauth_provider.data_endpoint:
-            data = self._auth_oauth_rpc(oauth_provider.data_endpoint, access_token)
+            data = super()._auth_oauth_rpc(oauth_provider.data_endpoint, access_token)
             validation.update(data)
         # unify subject key, pop all possible and get most sensible. When this
         # is reworked, BC should be dropped and only the `sub` key should be
@@ -75,7 +78,8 @@ class ResUsers(models.Model):
             None,
         )
         if not subject:
-            raise AccessDenied("Missing subject identity")
+            _logger.error("Access Denied: missing subject identity")
+            raise AccessDenied()
         validation["user_id"] = subject
 
         return validation
```

It seemed clearer this way:
* mimick what's done upstream in terms of methods
* rely on `super()._auth_oauth_rpc(oauth_provider.data_endpoint, access_token)` instead of doing a `requests.get` in the code removed below
* it's not clearer to parse the `id_token` before or separately.

I kept the two `_logger.error` + `raise AccessDenied` instead of copying upstream's, but I can certainly live with a closer copy.

Does this make it clearer?

I'm open to suggestions!

OdyX commented:
@sbidoul Force-pushed a rebase with the comment suggestion included in the first commit. (twice because I forgot about `pre-commit`).

sbidoul commented:
I'm not comfortable with replacing upstream method completely. Is there no other way? It is not clear to me why it is necessary for this PR but I have unfortunately no time to dig into this until October.

OdyX commented:
@sbidoul I managed to dive back into this code again. 

> I'm not comfortable with replacing upstream method completely. Is there no other way?

It turns out, there is. See https://github.com/OCA/server-auth/pull/682/commits/77c5c2d935906436d9c9f80853911ef4832b99b6 . In there, we just delete the duplicated code from `auth_oauth`, and merge the id_token and userinfo data into `validation`.

I'd be thrilled if we could merge this in 17.0, and I'd then be happy with forward porting these to 18.0 and 19.0 as needed.

(edited: sorry for the multiple force-pushes; tests turned out to be… relevant :-) )

OdyX committed:
[IMP] auth_oidc: clearer data_endpoint access

OdyX commented:
Note for future me; KeyCloak needs the system parameter `auth_oauth.authorization_header` set to 1, see https://github.com/odoo/odoo/blob/17.0/addons/auth_oauth/models/res_users.py#L33

OdyX commented:
> I'm not comfortable with replacing upstream method completely. Is there no other way?

It turns out, after more tests and a (yet) better understanding of the code, doing so is actually *worse* than using the upstream code: [the "Authorization: Bearer" is behind a system-level toggle](https://github.com/odoo/odoo/blob/17.0/addons/auth_oauth/models/res_users.py#L33), which makes things unnecessarily confusing and hard to debug.

I finally went with just a copy of the response checker in https://github.com/OCA/server-auth/commit/89a2a5f2f2159c9fcaa6b505f72fd5c5951659f4 (and I could live without that last commit).

OdyX commented:
I would really love this being merged. As far as I can tell, this is ready and just missing validation from @sbidoul .

So please remove this `stale` label.
